spec/models/feed_event_spec.rb

```diff
@@ -63,4 +63,129 @@
       end
     end
   end
+
+  describe "after_save .update_article_counters_and_scores" do
+    let!(:article) { create(:article) }
+    let!(:user1) { create(:user) }
+    let!(:user2) { create(:user) }
+
+    it "updates article counters and scores when a FeedEvent is saved" do
+      # Create some feed events to simulate behavior
+      create(:feed_event, category: "impression", article: article, user: user1)
+      create(:feed_event, category: "click", article: article, user: user1)
+      create(:feed_event, category: "reaction", article: article, user: user1)
+      create(:feed_event, category: "comment", article: article, user: user2)
+
+      # Trigger the after_save
+      create(:feed_event, category: "impression", article: article, user: user2)
+
+      # Reload the article to get the updated counters and scores
+      article.reload
+
+      expect(article.feed_success_score).to eq((1 + 5 + 10) / 2.0) # Calculated score
+      expect(article.feed_impressions_count).to eq(2) # Two impressions
+      expect(article.feed_clicks_count).to eq(1) # One click
+    end
+
+    it "returns early if article is nil" do
+      feed_event_without_article = build(:feed_event, category: "impression")
+      allow(feed_event_without_article).to receive(:update_article_counters_and_scores)
+      feed_event_without_article.save
+      expect(feed_event_without_article).not_to have_received(:update_article_counters_and_scores)
+    end
+
+    it "initializes counters and scores to zero when no events have occurred" do
+      # Trigger the after_save
+      create(:feed_event, category: "impression", article: article, user: user1)
+
+      article.reload
+
+      expect(article.feed_success_score).to eq(0.0)
+      expect(article.feed_impressions_count).to eq(1) # One impression
+      expect(article.feed_clicks_count).to eq(0) # Zero clicks
```

Comment:
```suggestion
      expect do
        create(:feed_event, category: "impression", article: article, user: user1)
      end.to not_change { article.reload.feed_success_score }
        .and not_change { article.reload.feed_clicks_count }
        .and change { article.reload.feed_impressions_count }.from(0).to(1)

```
To test callbacks effectively, you have to verify the value before and after, else it may get tampered by other test. This is the most succinct way. It has the benefit of self-explanation of what should be changing and how much.